File: src/main/java/me/kuku/yuq/controller/MenuController.java
package me.kuku.yuq.controller;

import com.IceCreamQAQ.Yu.annotation.Action;
import com.icecreamqaq.yuq.annotation.GroupController;
import me.kuku.yuq.controller.bilibili.BiliBiliController;
import me.kuku.yuq.controller.bilibili.BiliBiliLoginController;
import me.kuku.yuq.controller.manage.ManageAdminController;
import me.kuku.yuq.controller.manage.ManageNotController;
import me.kuku.yuq.controller.manage.ManageSuperAdminController;
import me.kuku.yuq.controller.netease.BindNeTeaseController;
import me.kuku.yuq.controller.netease.NeTeaseController;
import me.kuku.yuq.controller.qqlogin.BindQQController;
import me.kuku.yuq.controller.qqlogin.QQJobController;
import me.kuku.yuq.controller.qqlogin.QQLoginController;
import me.kuku.yuq.controller.qqlogin.QQQuickLoginController;
import me.kuku.yuq.controller.warframe.WarframeController;
import me.kuku.yuq.controller.weibo.WeiboController;
import me.kuku.yuq.controller.weibo.WeiboNotController;
import me.kuku.yuq.utils.BotUtils;

import java.util.List;

@GroupController
@SuppressWarnings("unused")
public class MenuController {

    @Action("help")
    public String help(){
        return parse("help", MenuController.class);
    }

    @Action("tool")
    public String tool(){
        return parse("tool", ToolController.class);
    }

    @Action("bilibili")
    public String bl(){
        return parse("bilibili", BiliBiliLoginController.class, BiliBiliController.class);
    }

    @Action("bot")
    public String bot(){
        return parse("bot", BotController.class);
    }

    @Action("manage")
    public String manage(){
        return parse("manage", ManageNotController.class, ManageSuperAdminController.class, ManageAdminController.class);
    }

    @Action("wy")
    public String wy(){
        return parse("网易", NeTeaseController.class, BindNeTeaseController.class);
    }

    @Action("qq")
    public String qq(){
        return parse("qq", QQLoginController.class, BindQQController.class, QQJobController.class, QQQuickLoginController.class);
    }

    @Action("setting")
    public String setting(){
        return parse("设置", SettingController.class);
    }

    @Action("wb")
    public String wb(){
        return parse("微博", WeiboNotController.class, WeiboController.class);
    }

    @Action("wf")
    public String wf() {
        return parse("wf",  WarframeController.class);
    }

    @Action("ark")
    public String ark(){
        return parse("ark", ArkNightsController.class);
    }

    @Action("菜单")
    public String menu(){
        return "https://api.kuku.me/menu";
    }


    private String parse(String name, Class<?>...classes){
        List<String> list = BotUtils.menu(classes);
        StringBuilder sb = new StringBuilder().append("╭┅═☆━━━━━━━").append(name).append("━━━┅╮").append("\n");
        for (int i = 0; i < list.size(); i+=2) {
            sb.append("||").append(list.get(i));
            if (i + 1 < list.size()){
                sb.append("   ").append(list.get(i + 1)).append("||");
            }else sb.append("||");
            if (i + 1 < list.size()){
                sb.append("\n");
            }
        }
        return sb.append("╰┅━━━━━━━━━━━━━★═┅╯").toString();
    }


}

File: src/main/kotlin/me/kuku/yuq/controller/MenuController.kt
@file:Suppress("unused")

package me.kuku.yuq.controller

import com.IceCreamQAQ.Yu.annotation.Action
import com.IceCreamQAQ.Yu.annotation.Synonym
import com.IceCreamQAQ.Yu.controller.router.NewRouterImpl
import com.IceCreamQAQ.Yu.di.YuContext
import com.icecreamqaq.yuq.annotation.GroupController
import com.icecreamqaq.yuq.controller.BotActionInvoker
import com.icecreamqaq.yuq.controller.BotReflectMethodInvoker
import me.kuku.yuq.utils.removeSuffixLine
import javax.inject.Inject

@GroupController
class MenuController @Inject constructor(private val yuContext: YuContext) {

    private fun getNewRouter(): NewRouterImpl{
        return yuContext.getBean("com.IceCreamQAQ.Yu.controller.router.NewRouter") as NewRouterImpl
    }

    @Action("help")
    fun help() = """
        tool
        bilibili
        bot
        manage
        motion
        wy
        qq
        setting
        wb
    """.trimIndent()

    @Action("tool")
    fun tool() = menu(ToolController::class.java)

    @Action("bilibili")
    fun bl() = menu(BiliBiliLoginController::class.java) + "\n" + menu(BiliBiliController::class.java)

    @Action("bot")
    fun bot() = menu(BotController::class.java)

    @Action("manage")
    fun manage() = menu(ManageNotController::class.java) + "\n" + menu(ManageOwnerController::class.java) +
            "\n" + menu(ManageAdminController::class.java)

    @Action("motion")
    fun motion() = menu(MotionController::class.java) + "\n" + menu(BindStepController::class.java)

    @Action("wy")
    fun wy() = menu(NeTeaseController::class.java) + "\n" + menu(BindNeTeaseController::class.java)

    @Action("qq")
    fun qq() = menu(QQController::class.java) + "\n" + menu(BindQQController::class.java) +
            "\n" + menu(QQJobController::class.java)

    @Action("setting")
    fun setting() = menu(SettingController::class.java)

    @Action("wb")
    fun wb() = menu(WeiboNotController::class.java) + "\n" + menu(WeiboController::class.java)

    fun menu(clazz: Class<*>): String{
        val sb = StringBuilder()
        val methods = clazz.methods
        for (method in methods){
            val action = method.getAnnotation(Action::class.java)
            if (action != null){
                sb.appendLine(action.value)
            }
            val synonym = method.getAnnotation(Synonym::class.java)
            if (synonym != null){
                val arr = synonym.value
                for (str in arr){
                    sb.appendLine(str)
                }
            }
        }
        return sb.removeSuffixLine().toString()
    }
}